### PR TITLE
Fix #784 Enable BasicAuth through Selenide proxy server

### LINKED ISSUE #784: Enable BasicAuth through Selenide proxy server
## The problem
Currently it's only possible to bypass BasicAuth by adding username+password in URL. And this is deprecated now (at least in Chrome). We could add BasicAuth header using Selenide's built-in proxy server.

### PR DESCRIPTION
**Pull request by rosolko**
## Proposed changes
Fix #784 issue. Enable BasicAuth through Selenide proxy server

## Checklist
- [x] Checkstyle and unit tests pass locally with my changes by running `gradle check chrome htmlunit` command
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

### CODE REVIEW

File: src/main/java/com/codeborne/selenide/Selenide.java [hunk 1]
```diff
@@ -101,6 +104,35 @@ public static void open(String relativeOrAbsoluteUrl, String domain, String logi
...
+   * @param authenticationType
+   * @param login
+   * @param password
+   *
+   * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Proxy-Authorization">Web HTTP reference</a>
+   * @see AuthenticationType
+   */
+  public static void open(final String relativeOrAbsoluteUrl, final AuthenticationType authenticationType,
+                          final String login, final String password) {
+    getAndCheckWebDriver();
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/Selenide.java]: Here is too much logic for `Selenide` class. All this stuff should be in `SelenideProxyServer` and/or `navigator.open`. In this case we would not need to increase `ClassFanOutComplexity` in checkstyle.xml

File: src/main/java/com/codeborne/selenide/Selenide.java [hunk 2]
```diff
@@ -101,6 +104,35 @@ public static void open(String relativeOrAbsoluteUrl, String domain, String logi
...
+    final BrowserMobProxy proxy = getSelenideProxy().getProxy();
+    final Credentials credentials = new Credentials(login, password);
+    final String authorization = String.format("%s %s", authenticationType.getValue(), credentials.encode());
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/Selenide.java]: I personally do not like `final` near every local variable and method parameter. It's a noise.

File: src/main/java/com/codeborne/selenide/Selenide.java [hunk 3]
```diff
@@ -101,6 +104,35 @@ public static void open(String relativeOrAbsoluteUrl, String domain, String logi
...
+   * @param login
+   * @param password
+   *
+   * @see <a href="https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Proxy-Authorization">Web HTTP reference</a>
+   * @see AuthenticationType
+   */
+  public static void open(final String relativeOrAbsoluteUrl, final AuthenticationType authenticationType,
+                          final String login, final String password) {
+    getAndCheckWebDriver();
+    final BrowserMobProxy proxy = getSelenideProxy().getProxy();
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/Selenide.java]: It's not a good idea to use BrowserMobProxy directly. Much better would be to add `RequestFilter` using `getSelenideProxy().addRequestFilter()`. This filter could add required headers.

**Comment by rosolko** [comment]
@asolntsev Done.

File: src/main/java/com/codeborne/selenide/impl/Navigator.java [hunk 4]
```diff
@@ -39,6 +43,19 @@ public void open(URL url, String domain, String login, String password) {
...
   }
 
+  public void open(String relativeOrAbsoluteUrl, AuthenticationType authenticationType, String login, String password) {
+    getAndCheckWebDriver();
+    Credentials credentials = new Credentials(login, password);
+    String authorization = String.format("%s %s", authenticationType.getValue(), credentials.encode());
+    getSelenideProxy().addRequestFilter("headers.request", (request, contents, messageInfo) -> {
+      final HttpHeaders headers = request.headers();
+      headers.add("Authorization", authorization);
+      headers.add("Proxy-Authorization", authorization);
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/impl/Navigator.java]: @rosolko Isn't it a problem that we add this filter, but never remove?
It means that it will add `Authorization` header to ALL subsequent requests. Is it ok? Or we should remove this filter?

File: src/main/java/com/codeborne/selenide/Credentials.java [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+  public String encode() {
+    final byte[] credentialsBytes = combine().getBytes(UTF_8);
+    return Base64.getEncoder().encodeToString(credentialsBytes);
+  }
+
+  /**
+   * Combine credentials with a colon (login:password).
+   *
+   * @return combined credentials
+   */
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/Credentials.java]: Do we really need javadoc for private methods?

File: src/main/java/com/codeborne/selenide/Credentials.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
+package com.codeborne.selenide;
+
+import java.util.Base64;
+
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+public final class Credentials {
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/Credentials.java]: I would move this class to package `com.codeborne.selenide.impl` and make package-private. It's only used in `Navigator`.

File: src/main/java/com/codeborne/selenide/AuthenticationType.java [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+ */
+public enum AuthenticationType {
+  BASIC("Basic"),
+  BEARER("Bearer"),
+  DIGEST("Digest"),
+  HOBA("HOBA"),
+  MUTUAL("Mutual"),
+  AWS4_HMAC_SHA256("AWS4-HMAC-SHA256");
+
+  private String value;
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/AuthenticationType.java]: I would add `private final` here.

File: src/main/java/com/codeborne/selenide/AuthenticationType.java [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+
+  AuthenticationType(final String value) {
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/AuthenticationType.java]: I would remove `final` because it does't add any value.

File: src/main/java/com/codeborne/selenide/Selenide.java [hunk 9]
```diff
@@ -101,6 +101,28 @@ public static void open(String relativeOrAbsoluteUrl, String domain, String logi
...
+   * Open browser and pass authentication using build-in proxy.
+   * <p>
+   * A common authenticationType is "Basic". See Web HTTP reference for other types.
+   * <p>
+   * {@code Configuration.fileDownload == Configuration.FileDownloadMode.PROXY;} - is required configuration.
+   *
+   * @param relativeOrAbsoluteUrl
+   * @param authenticationType
+   * @param login
+   * @param password
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/Selenide.java]: I would remove all useless lines from javadoc, like `@param login` without description.

File: src/main/java/com/codeborne/selenide/Selenide.java [hunk 10]
```diff
@@ -101,6 +101,28 @@ public static void open(String relativeOrAbsoluteUrl, String domain, String logi
...
 
+  /**
+   *
+   * The main starting point in your tests.
+   * <p>
+   * Open browser and pass authentication using build-in proxy.
+   * <p>
+   * A common authenticationType is "Basic". See Web HTTP reference for other types.
+   * <p>
+   * {@code Configuration.fileDownload == Configuration.FileDownloadMode.PROXY;} - is required configuration.
```
Comment (asolntsev) [inline src/main/java/com/codeborne/selenide/Selenide.java]: I would rephrase it to something like "this method can only work if `{@code Configuration.fileDownload == Configuration.FileDownloadMode.PROXY;}`"

File: src/main/java/com/codeborne/selenide/impl/Navigator.java (hunk 4 above)
Comment (rosolko) [reply src/main/java/com/codeborne/selenide/impl/Navigator.java]: @asolntsev 
1. Previously you sad that we shouldn't call internal `getProxy()` and call only top level selenide proxy `getSelenideProxy()`
2. `autoAuthorization` method supports only `BASIC` auth type
3. If we remove filter isn't application logout user after another action in application?
4. If remove - how can we remove filters from selenide proxy? I see only add functionality.

File: src/main/java/com/codeborne/selenide/Credentials.java (hunk 5 above)
Comment (rosolko) [reply src/main/java/com/codeborne/selenide/Credentials.java]: @asolntsev 
Yes we do.
The main point thats this object is public and method like this (I mean do some combining logic with special format) is necessary to de described in any way.

File: src/main/java/com/codeborne/selenide/Credentials.java (hunk 6 above)
Comment (rosolko) [reply src/main/java/com/codeborne/selenide/Credentials.java]: @asolntsev 
I consider to move this class to top level because I think about extra `open` method with signatures containing `credentials` instead of `login` and `password` pair.
What do you thinks?

File: src/main/java/com/codeborne/selenide/AuthenticationType.java (hunk 7 above)
Comment (rosolko) [reply src/main/java/com/codeborne/selenide/AuthenticationType.java]: @asolntsev Good catch!

File: src/main/java/com/codeborne/selenide/Selenide.java (hunk 9 above)
Comment (rosolko) [reply src/main/java/com/codeborne/selenide/Selenide.java]: @asolntsev 
We already had such empty javadocs lines for old `open(relativeOrAbsoluteUrl, domain, login, password)` method. 
I do it in one style during new implementation.
In this case we had to reformat all methods related to this point.

File: src/main/java/com/codeborne/selenide/Selenide.java (hunk 10 above)
Comment (rosolko) [reply src/main/java/com/codeborne/selenide/Selenide.java]: @asolntsev Understood. Thanks!

File: src/main/java/com/codeborne/selenide/impl/Navigator.java (hunk 4 above)
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/impl/Navigator.java]: 1. Agree
2. Agree
3. I don't know exactly if logout happens. Probably. But anyway, we need to be able to somehow disable filter (when user calls `open` without basicauth) or reconfigure it (when user calls `open` with another username/password).
4. Right, it seems that we cannot remove filter from BMP. So, we can add filter only once, and enable/disable or configure existing filter.

File: src/main/java/com/codeborne/selenide/Credentials.java (hunk 5 above)
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/Credentials.java]: Let's make it public when it's really needed. Currently it's not used, so there is no reason to make it public.

File: src/main/java/com/codeborne/selenide/Credentials.java (hunk 6 above)
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/Credentials.java]: Yes, one parameter `Credentials` is better than two Strings "login"/"password".

File: src/main/java/com/codeborne/selenide/Selenide.java (hunk 9 above)
Comment (asolntsev) [reply src/main/java/com/codeborne/selenide/Selenide.java]: I agree, we can cleanup javadoc later.

File: src/main/java/com/codeborne/selenide/Credentials.java (hunk 5 above)
Comment (rosolko) [reply src/main/java/com/codeborne/selenide/Credentials.java]: Agree. Just remove javadoc since it's specially private method.

File: src/main/java/com/codeborne/selenide/Credentials.java (hunk 6 above)
Comment (rosolko) [reply src/main/java/com/codeborne/selenide/Credentials.java]: Or we can just overload method and give ability to do open with `Credentials` and `login/password` pair?
